### ISSUE TITLE
Problematic handling of Python slices

### ISSUE THREAD

**Issue by tommadams** [google/jax#3708]
The [documentation](https://jax.readthedocs.io/en/latest/jax.html?highlight=jit#jax.jit) for `jax.jit` states:

>Positional arguments indicated by static_argnums can be anything at all, provided they are hashable and have an equality operation defined.

However, slice objects are not hashable but can be passed as static_args without error:

```py
@functools.partial(jax.jit, static_argnums=1)
def foo(arr, s):
  return arr[s]

a = jax.numpy.array([1, 2, 3])
foo(a, slice(1, 2))
```

The current handling of slice static_args is problematic because JAX appears to be using the slice's object ID as its cache key. This means that passing different but equal valued slice instances will cause the function to be recompiled each time, e.g.:

```py
# foo will be compiled twice
foo(a, slice(1, 2))
foo(a, slice(1, 2))
```

**Comment by mattjj** [comment]
Thanks for finding this!

One change we've discussed (and I _think_ had unanimous support) is only to allow hashable values to be passed as static arguments. That is, the behavior of handling even unhashable objects by hashing based on id seems too surprising (i.e. leading to surprising recompiles). Given that change, we'd at least get a loud error here.

Additionally, we could likely handle slices as containers (i.e. pytrees). Then this case would work as expected, getting cache hits for equal slice values but recompiling for distinct slice values. I think we attempted that before and ran into some complexity which wasn't insurmountable but required more effort than we wanted to invest. Perhaps we should try that again!

WDYT?

**Comment by tommadams** [comment]
I'm 100% in favour of only allowing hashable values as static args. When I first ran into this issue, I assumed it was a bug in our own code (we have wrappers around jax.jit) because I found the behaviour so surprising.

Handling slices as pytrees would be nice to have but not a high priority for us. In our codebase at least I'm only aware of a single case where we were passing a slice to a jitted function, and refactoring the code to avoid that was a small localised change.
